File: sdks/python/ans/integrations/langgraph.py
"""LangGraph integration for ANS."""

from typing import Any, Callable

try:
    from langgraph.graph import StateGraph
    _langgraph_available = True
except ImportError:
    StateGraph = None
    _langgraph_available = False

from ..client import ANSClient
from ..middleware import ANSMiddleware


def ans_node(
    func: Callable,
    client: ANSClient,
    agent_id: str,
    parent_agent_id: str = "",
    action_type: str = "agent.delegate",
) -> Callable:
    """
    Wrap a LangGraph node function with ANS tracing.
    
    Args:
        func: Node function to wrap
        client: ANS client
        agent_id: Agent ID
        parent_agent_id: Parent agent ID (if sub-agent)
        action_type: Action type for receipts
    
    Example:
        from ans.integrations.langgraph import ans_node
        
        def my_node(state):
            # ... node logic
            return state
        
        traced_node = ans_node(my_node, client, agent_id="ans_xyz")
        
        graph = StateGraph(...)
        graph.add_node("my_node", traced_node)
    """
    middleware = ANSMiddleware(client, agent_id, parent_agent_id)

    return middleware.wrap(
        func,
        action_type=action_type,
        payload_summary_fn=lambda state: f"Node: {func.__name__}",
    )


class ANSGraph:
    """
    Wrapper for StateGraph that automatically traces all node executions.
    
    Example:
        from ans.integrations.langgraph import ANSGraph
        
        graph = ANSGraph(client, agent_id="ans_xyz")
        graph.add_node("node1", node1_func)
        graph.add_node("node2", node2_func)
        graph.add_edge("node1", "node2")
        
        app = graph.compile()
        app.invoke({"input": "..."})
    """

    def __init__(
        self,
        client: ANSClient,
        agent_id: str,
        parent_agent_id: str = "",
        state_schema: Any = None,
    ):
        if not _langgraph_available:
            raise ImportError("LangGraph is not installed. Install with: pip install ans-sdk[langgraph]")
        self.client = client
        self.agent_id = agent_id
        self.parent_agent_id = parent_agent_id
        self.graph = StateGraph(state_schema) if state_schema else StateGraph(dict)
        self.middleware = ANSMiddleware(client, agent_id, parent_agent_id)

    def add_node(self, name: str, func: Callable) -> None:
        """Add a node with automatic ANS tracing."""
        wrapped = self.middleware.wrap(
            func,
            action_type="agent.delegate",
            payload_summary_fn=lambda state: f"Node: {name}",
        )
        self.graph.add_node(name, wrapped)

    def add_edge(self, start: str, end: str) -> None:
        """Add an edge between nodes."""
        self.graph.add_edge(start, end)

    def add_conditional_edges(self, source: str, condition: Callable, mapping: dict) -> None:
        """Add conditional edges."""
        self.graph.add_conditional_edges(source, condition, mapping)

    def set_entry_point(self, node: str) -> None:
        """Set the entry point of the graph."""
        self.graph.set_entry_point(node)

    def set_finish_point(self, node: str) -> None:
        """Set the finish point of the graph."""
        self.graph.set_finish_point(node)

    def compile(self) -> Any:
        """Compile the graph."""
        return self.graph.compile()

File: sdks/typescript/src/integrations/langgraph.ts
/**
 * LangGraph integration for ANS.
 *
 * Provides ANSGraph — a drop-in wrapper for StateGraph that automatically
 * traces every node execution with ANS receipts.
 *
 * Usage:
 *   import { ANSGraph } from "ans-sdk/integrations/langgraph";
 *   import { Client } from "ans-sdk";
 *
 *   const graph = new ANSGraph({ agentId: "ans_xyz123" });
 *   graph.addNode("my_node", (state) => ({ ...state, processed: true }));
 *   graph.addEdge("__start__", "my_node");
 *   const app = graph.compile();
 *   const result = await app.invoke({ input: "hello" });
 *
 * SPDX-License-Identifier: Apache-2.0
 */
import { Client, ANSError, ActionType } from "../client";
import { wrap } from "../wrap";

let StateGraph: any;
try {
  StateGraph = require("@langchain/langgraph").StateGraph;
} catch {
  throw new Error(
    "@langchain/langgraph is not installed. Install it with: npm install @langchain/langgraph"
  );
}

export function ansNode(
  func: (...args: unknown[]) => unknown,
  opts: { agentId: string; client?: Client; silent?: boolean; actionType?: ActionType }
): (...args: unknown[]) => unknown {
  return wrap(func, {
    agentId: opts.agentId,
    actionType: opts.actionType || ("agent.delegate" as ActionType),
    client: opts.client,
    silent: opts.silent,
  });
}

export class ANSGraph {
  private client: Client;
  private agentId: string;
  private silent: boolean;
  private graph: any;

  constructor(opts: { agentId: string; client?: Client; silent?: boolean }) {
    this.client = opts.client || new Client();
    this.agentId = opts.agentId;
    this.silent = opts.silent !== false;
    this.graph = new StateGraph({ channels: {} });
  }

  addNode(name: string, func: (...args: unknown[]) => unknown): void {
    const wrapped = wrap(func, {
      agentId: this.agentId,
      actionType: "agent.delegate" as ActionType,
      client: this.client,
      silent: this.silent,
    });
    this.graph.addNode(name, wrapped);
  }

  addEdge(start: string, end: string): void {
    this.graph.addEdge(start, end);
  }

  addConditionalEdges(source: string, condition: (...args: unknown[]) => string, mapping: Record<string, string>): void {
    this.graph.addConditionalEdges(source, condition, mapping);
  }

  setEntryPoint(node: string): void {
    this.graph.setEntryPoint(node);
  }

  setFinishPoint(node: string): void {
    this.graph.setFinishPoint(node);
  }

  compile(): any {
    return this.graph.compile();
  }
}
